### ISSUE TITLE
Crash: object over release

### ISSUE THREAD

**Issue by 10Macit** [GetStream/stream-chat-swift#2782]
We've integrated the StreamChatSwiftUI SDK into our iOS app and rolled out the update 2 weeks ago, but the chat features still disabled through remote configuration as we continue thorough testing.

However, we encountered an issue where two users experienced crashes related to the Stream SDK in our production environment. These crashes occurred only once for these two users, and unfortunately, we haven't been able to replicate the issue during our testing.

## GetStream Environment
**GetStream Chat version: 4.35.1
**GetStream Chat frameworks:** StreamChat, StreamChatSwiftUI
**iOS version:16.6.1
**Swift version:5
**Xcode version:14.3.1
**Device:iPhone 11- iPhone 11 Pro Max

## Additional context
**Keys:** 
crash_info_entry_0 : API MISUSE: Over-release of an object 
**Stack trace:**
Crashed: com.stream.api-client (QOS: UNSPECIFIED)
0  libdispatch.dylib              0x2984 _os_object_release + 76
1  StreamChat                     0x42b78 AuthenticationRepository.deinit + 30 (AuthenticationRepository.swift:30)
2  StreamChat                     0x42c34 AuthenticationRepository.__deallocating_deinit + 30 (AuthenticationRepository.swift:30)
3  libswiftCore.dylib             0x3bd618 _swift_release_dealloc + 56
4  libswiftCore.dylib             0x3be43c bool swift::RefCounts<swift::RefCountBitsT<(swift::RefCountInlinedness)1> >::doDecrementSlow<(swift::PerformDeinit)1>(swift::RefCountBitsT<(swift::RefCountInlinedness)1>, unsigned int) + 132
5  StreamChat                     0xe6a7c ChatClient.authenticationRepository.getter + 97 (ChatClient.swift:97)
6  StreamChat                     0x2564bc specialized DefaultRequestEncoder.addAuthorizationHeader<A>(request:endpoint:completion:) + 1076
7  StreamChat                     0x27ca9c specialized DefaultRequestEncoder.encodeRequest<A>(for:completion:) + 114 (RequestEncoder.swift:114)
8  StreamChat                     0x2684dc specialized closure GetStream/stream-chat-swiftui#1 in APIClient.operation<A>(endpoint:isRecoveryOperation:completion:) + 218 (APIClient.swift:218)
9  StreamChat                     0x182dc AsyncOperation.start() + 37 (AsyncOperation.swift:37)
10 StreamChat                     0x1831c @objc AsyncOperation.start() + 28 (<compiler-generated>:28)
11 Foundation                     0x3c1ec __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 16
12 Foundation                     0x4187c __NSOQSchedule_f + 172
13 libdispatch.dylib              0x12fc4 _dispatch_block_async_invoke2 + 148
14 libdispatch.dylib              0x3eac _dispatch_client_callout + 20
15 libdispatch.dylib              0x7330 _dispatch_continuation_pop + 504
16 libdispatch.dylib              0x6998 _dispatch_async_redirect_invoke + 584
17 libdispatch.dylib              0x15944 _dispatch_root_queue_drain + 396
18 libdispatch.dylib              0x16158 _dispatch_worker_thread2 + 164
19 libsystem_pthread.dylib        0xda0 _pthread_wqthread + 228
20 libsystem_pthread.dylib        0xb7c start_wqthread + 8

**Integration**
class AppDelegate: NSObject, UIApplicationDelegate {
    
    var streamChat: StreamChat?
    
    var chatClient: ChatClient = {
        var config = ChatClientConfig(apiKey: .init(AppConfiguration.chatKey))
        config.isLocalStorageEnabled = true
        config.applicationGroupIdentifier = AppConfiguration.applicationGroupIdentifier
        let client = ChatClient(config: config)
        return client
    }()
    
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
        let config = MessageListConfig(updateChannelsFromMessageList: true, handleTabBarVisibility: false)
        let utils = Utils(messageListConfig: config)
        streamChat = StreamChat(chatClient: chatClient, utils: utils)
        #if DEBUG
        LogConfig.level = .info
        #endif
    }   
}

**Comment by martinmitrevski** [comment]
Thanks for reporting. Will transfer this to our `StreamChat` repo, since the issue is probably there. Anything else that you can share about how this occurs? Can you share your logout logic?

**Comment by 10Macit** [comment]
I found a one issue from our side, accessing to channel while `connectUser` is not called. But this issue is also not crashing for other users or not crashing again at the moment. All the other chat usage like connectUser etc. disabled with remote configuration.
```
class CollaborationCellViewModel: ObservableObject {
    @Injected(\.chatClient) var chatClient

    var channelController: ChatChannelController!
    @Published var unreadCount: Int = 0

    func getChatMessageCount(cid: String) {
        guard let channel = try? ChannelId(cid: cid) else { return }
        if channelController == nil {
            channelController = try chatClient.channelController(
                for: channel,
                messageOrdering: .topToBottom
            )
            channelController.synchronize { [weak self] error in
                if error != nil {
                }
            }
            unreadCount = channelController?.channel?.unreadCount.messages ?? 0
        } else {
            unreadCount = channelController.channel?.unreadCount.messages ?? 0
        }
    }
}
```
Logout logic
```
    func logout() {
        chatClient.disconnect {
            print("disconnected")
        }
        Authentication.shared.logoutUser()
    }
```

**Comment by nuno-vieira** [comment]
Hi @10Macit,

You should make sure that you call connectUser and wait for a successful connection before using any controller. 

When you logout the user, you should also use `chatClient.logout()` and not `chatClient.disconnect()` read more about this: `https://getstream.io/chat/docs/sdk/ios/uikit/getting-started/#disconnect--logout`

It is important as well that you wait for the `logout()` completion block and do not ignore it. This will take milliseconds it won't block the UI, no problem. 

These are all best practices right now until we can change how our login/logout flow works, and make it more safe.

**Comment by nuno-vieira** [comment]
Either way, the new release that is coming out next week should make this more robust 👍

**Comment by nuno-vieira** [comment]
Hi!

This issue should have been resolved on the new [4.37.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.37.0) Release.

Please re-open the issue if you still have crash reports after updating.

Best,
Nuno
